pints/_optimisers/_hager_zhang_line_search.py

```diff
@@ -0,0 +1,899 @@
+#
+# Fixed learning-rate gradient descent.
+#
+# This file is part of PINTS (https://github.com/pints-team/pints/) which is
+# released under the BSD 3-clause license. See accompanying LICENSE.md for
+# copyright notice and full license details.
+#
+from __future__ import absolute_import, division
+from __future__ import print_function, unicode_literals
+
+import pints
+import numpy as np
+from numpy.linalg import norm
+
+
+class HagerZhang(pints.Optimiser):
+    """
+    Gradient-descent method with a fixed learning rate.
+    """
+
+    def __init__(self, x0, sigma0=None, boundaries=None):
+        super(HagerZhang, self).__init__(x0, sigma0, boundaries)
+
+        # Set optimiser state
+        self._running = False
+        self._ready_for_tell = False
+
+        # Best solution found
+        self._xbest = self._x0
+        self._fbest = float('inf')
+        self._dfdx_best = [float('inf'), float('inf')]
+
+        # Number of iterations run
+        self._iterations = 0
+
+        # Parameters for wolfe conditions on line search
+
+        # As c1 approaches 0 and c2 approaches 1, the line search
+        # terminates more quickly.
+        self._c1 = 1E-4  # Parameter for Armijo condition rule, 0 < c1 < 0.5
```

Comment:
I couldn't find the Armijo rule in the paper? Could you point to where this is from? I'm guessing this is "delta" on page 184 of Hager and Zhang?

Reply:
Also, where the default value was taken from?